lib/block_view.go

```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
+
+func (bav *UtxoView) _connectAcceptNFTTransfer(
+	txn *MsgBitCloutTxn, txHash *BlockHash, blockHeight uint32, verifySignatures bool) (
+	_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
+
+	if blockHeight < NFTTransferOrBurnBlockHeight {
+		return 0, 0, nil, RuleErrorAcceptNFTTranserBeforeBlockHeight
+	}
+
+	// Check that the transaction has the right TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeAcceptNFTTransfer {
+		return 0, 0, nil, fmt.Errorf("_connectAcceptNFTTransfer: called with bad TxnType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+	txMeta := txn.TxnMeta.(*AcceptNFTTransferMetadata)
+
+	// Verify the NFT entry exists.
+	nftKey := MakeNFTKey(txMeta.NFTPostHash, txMeta.SerialNumber)
+	prevNFTEntry := bav.GetNFTEntryForNFTKey(&nftKey)
+	if prevNFTEntry == nil || prevNFTEntry.isDeleted {
+		return 0, 0, nil, RuleErrorCannotAcceptTransferOfNonExistentNFT
+	}
+
+	// Verify that the updater is the owner of the NFT.
+	updaterPKID := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if updaterPKID == nil || updaterPKID.isDeleted {
+		return 0, 0, nil, fmt.Errorf("_connectAcceptNFTTransfer: non-existent updaterPKID: %s",
+			PkToString(txn.PublicKey, bav.Params))
+	}
+	if !reflect.DeepEqual(prevNFTEntry.OwnerPKID, updaterPKID.PKID) {
+		return 0, 0, nil, RuleErrorAcceptNFTTransferByNonOwner
+	}
+
+	// Verify that the NFT is actually pending.
+	if !prevNFTEntry.IsPending {
+		return 0, 0, nil, RuleErrorAcceptNFTTransferForNonPendingNFT
+	}
+
+	// Sanity check that the NFT entry is not for sale.
+	if prevNFTEntry.IsForSale {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectAcceptNFTTransfer: attempted to accept NFT transfer of NFT that is for "+
+				"sale. This should never happen; txMeta %v.", txMeta)
+	}
+
+	// Sanity check that the NFT entry is correct.
+	if !reflect.DeepEqual(prevNFTEntry.NFTPostHash, txMeta.NFTPostHash) ||
+		!reflect.DeepEqual(prevNFTEntry.SerialNumber, txMeta.SerialNumber) {
+		return 0, 0, nil, fmt.Errorf("_connectAcceptNFTTransfer: prevNFTEntry %v is "+
+			"inconsistent with txMeta %v; this should never happen.", prevNFTEntry, txMeta)
+	}
+
+	// Connect basic txn to get the total input and the total output without
+	// considering the transaction metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectAcceptNFTTransfer: ")
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorAcceptNFTTransferRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
+		// public key.
+	}
+
+	// Now we are ready to transfer the NFT.
+
+	// Create the updated NFTEntry (everything the same except for IsPending) and set it.
+	newNFTEntry := &NFTEntry{}
```

Comment:
I think this works but prefer:

```
// Make a copy of the previous NFT entry
newNFTEntry := *prevNFTEntry
// Set the NFT to no longer be pending
newNFTEntry.IsPending = false
// Grab the pointer here.
bav._setNFTEntryMappings(&newNFTEntry)
```

Reply:
Appears to be a duplicate comment.  Resolving.